src/core/border_router/routing_manager.cpp

```diff
@@ -578,11 +578,25 @@ void RoutingManager::EvaluateOnLinkPrefix(void)
             LogInfo("Start advertising on-link prefix %s on %s", mLocalOnLinkPrefix.ToString().AsCString(),
                     mInfraIf.ToString().AsCString());
 
-            // Call `InvalidateDiscoveredPrefixes()` after setting
-            // `mIsAdvertisingLocalOnLinkPrefix` to remove on-link
-            // prefix in case it was discovered and included in
-            // `mDiscoveredPrefixes` list earlier.
-            InvalidateDiscoveredPrefixes();
+            // We go through `mDiscoveredPrefixes` list to check if the
+            // local on-link prefix was previously discovered and
+            // included in the list and if so we remove it from list.
+            //
+            // Note that `UpdateDiscoveredOnLinkPrefix()` will also
+            // check and not add local on-link prefix in the discovered
+            // prefix list while we are advertising the local on-link
+            // prefix.
+
+            for (ExternalPrefix &prefix : mDiscoveredPrefixes)
```

Comment:
How do we make sure removing while traversing works without issue?

Reply:
We `break` as soon as we find a match and then we remove it.

One idea is that we may later add an API `Array::RemoveMatching()` that will be common and do the same.

Reply:
That should be fine.

Thanks for the tip.